src/dmd/lexer.d

```diff
@@ -300,62 +403,60 @@ class Lexer
         return diagnosticReporter.errorCount > 0;
     }
 
-    /// Returns: a newly allocated `Token`.
-    Token* allocateToken() pure nothrow @safe
+    private ref Token getByIndex(size_t index)
     {
-        if (tokenFreelist)
+        while (index >= tokensScanned)
         {
-            Token* t = tokenFreelist;
-            tokenFreelist = t.next;
-            t.next = null;
-            return t;
+            Token* current = &tokenPool.getByIndex(tokensScanned);
+            scan(current);
+            tokensScanned++;
         }
-        return new Token();
+        return tokenPool.getByIndex(index);
     }
 
-    /// Frees the given token by returning it to the freelist.
-    private void releaseToken(Token* token) pure nothrow @nogc @safe
-    {
-        if (mem.isGCEnabled)
-            *token = Token.init;
-        token.next = tokenFreelist;
-        tokenFreelist = token;
-    }
 
     final TOK nextToken()
     {
         prevloc = token.loc;
-        if (token.next)
-        {
-            Token* t = token.next;
-            memcpy(&token, t, Token.sizeof);
-            releaseToken(t);
-        }
-        else
-        {
-            scan(&token);
-        }
-        //printf(token.toChars());
+        token = getByIndex(nextTokenIndex++);
         return token.value;
+
+    }
+
+    final size_t currentPosition()
+    {
+        return nextTokenIndex;
+    }
+
+    final void seekTo(size_t index)
+    {
+        token = getByIndex(index);
+        nextTokenIndex = index + 1;
+    }
+
+    ///make a range containing only this.token
+    final TokenRange makeRangeFromHere()
```

Comment:
Perhaps `currentTokenRange` would be a better name?

Reply:
I like the idea of opslice as there's a couple places where I advance it immediately.  My worry is that it might be a little confusing what the indices mean, but I supposed making them relative to the current token makes sense.  Good suggestion!

Reply:
You don't need any indexes. Just declare it as `final TokenRange opSlice()`.